### ISSUE TITLE
Resolution should try the next one if the first satisfied one fails

### ISSUE THREAD

**Issue by CNSeniorious000** [pyodide/micropip#113]
## Use case:

`haystack-ai==2.2` requires `lxml`, which don't have a wheel on pyodide. But `haystack<2.2` don't requires `lxml`.

Resolution should try the next one if the first satisfied one fails. In the example above, when resolving `lxml` raises `ValueError`, so haystask 2.2 fails, then it should try haystask 2.1.2 instead of bubbling the ValueError.

## Reproduction:

```py
from micropip import install
await install("haystack-ai")  # this will fail because lxml
await install("haystack-ai<2.2")  # this will success
```

**Comment by hoodmane** [comment]
See #112. I don't think we want to be a backtracking resolver by default because it's an np-hard problem to solve a dep set. Particularly until we start using [PEP 658](https://peps.python.org/pep-0658/), this could lead to downloading humongous amounts of stuff while looking for the right solve. It's not ideal for people to be using micropip at runtime anyways, so maybe this would be for creating a lock file ahead of time and then deploying the solve. I was hoping to use pip for this, see #107.

**Comment by CNSeniorious000** [comment]
I understand. So this could be discussed later when we started using PEP 658.

> It's not ideal for people to be using micropip at runtime anyways, so maybe this would be for creating a lock file ahead of time and then deploying the solve.

Yes of course solving the dependency graph ahead of time and only once is better. But there exists cases when runtime resolution is necessary, such as:

- trying if a package is compatible in the playground
- in a 3-party hosted jupyterlite deployment, running `%pip install ...` some package that not in the lock file
- LLM code interpreter

So I think a backtracking resolver is necessary but it can be defaults to false, like `keep_going`.

And I think it is easy to implement, by just making the `ver` a list or an iterator or satisfied versions [here](https://github.com/pyodide/micropip/blob/7ad29637186baad68df33c3c92e744dd45da7c12/micropip/transaction.py#L194) can achieve the my ideal result.

**Comment by hoodmane** [comment]
> `lxml`, which don't have a wheel on pyodide

As an aside, `lxml` is included in Pyodide: https://github.com/pyodide/pyodide/blob/main/packages/lxml/meta.yaml

**Comment by hoodmane** [comment]
I think we should do #112 and then add a backend that uses resolvelib to do a proper solve.
https://pypi.org/project/resolvelib/

**Comment by ryanking13** [comment]
Yep, I would like to refactor the micropip (#112) before introducing large changes like `resolvelib` or `installer` or PEP 658 (FYI, there is a draft PR for PEP 658 #90). It is not my top priority so if anyone is interested, it's okay to work on this.
